### ISSUE TITLE
Excluding OS Specific Dependencies Fails with ModuleNotFoundError

### ISSUE THREAD

**Issue by danyeaw** [jwbargsten/pytest-archon#26]
In https://github.com/gaphor/gaphor/pull/2045, I tried adding a dependency just for macOS:
`pyobjc = {version = "^9.0.1", markers = "sys_platform == 'darwin'"}`

We are then making use of the import in our i18n module:

```
    if sys.platform == "darwin":
        from Cocoa import NSUserDefaults

        defaults = NSUserDefaults.standardUserDefaults()
        langs = defaults.objectForKey_("AppleLanguages")
        if language := langs.objectAtIndex_(0):
            return language
```

Then we are testing it with pytest-archon like:

```
def test_core_packages():
    (
        archrule("gaphor.core does not depend on the rest of the system")
        .match("gaphor.core*")
        .exclude("*.tests.*")
        .may_import(*GAPHOR_CORE)
        .should_not_import("gaphor*")
        .should_not_import(*UI_LIBRARIES)
        .should_not_import("Cocoa")
        .check(gaphor, skip_type_checking=True)
    )
```

However this results in:

```
    def test_core_packages():
        (
            archrule("gaphor.core does not depend on the rest of the system")
            .match("gaphor.core*")
            .exclude("*.tests.*")
            .may_import(*GAPHOR_CORE)
            .should_not_import("gaphor*")
            .should_not_import(*UI_LIBRARIES)
            .should_not_import("Cocoa")
>           .check(gaphor, skip_type_checking=True)
        )

tests\test_architecture.py:37:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  
.venv\Lib\site-packages\pytest_archon\rule.py:176: in check
    all_imports = collect_imports(
.venv\Lib\site-packages\pytest_archon\collect.py:35: in collect_imports
    for name, imports in collect_imports_from_path(package_dir(package), package, walker):
.venv\Lib\site-packages\pytest_archon\collect.py:85: in collect_imports_from_path
    return frozenset(_collect(py_file) for py_file in Path(path).glob("**/*.py"))
.venv\Lib\site-packages\pytest_archon\collect.py:85: in <genexpr>
    return frozenset(_collect(py_file) for py_file in Path(path).glob("**/*.py"))
.venv\Lib\site-packages\pytest_archon\collect.py:83: in _collect
    return module_name, frozenset(imports)
.venv\Lib\site-packages\pytest_archon\collect.py:113: in extract_imports_ast
    yield resolve_module_or_object_by_path(fqname)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  

fqname = 'Cocoa.NSUserDefaults'

    def resolve_module_or_object_by_path(fqname: str) -> str:
        if "." not in fqname:
            return fqname

        if fqname in core_modules():
            return fqname

        parts = fqname.split(".")
        head = parts[0]
        parent_name = ".".join(parts[:-1])

        # shortcut to deal with e.g. from __future__ import annotations
        if parent_name in core_modules():
            return parent_name

        # taken from importlib.util.find_spec
        if fqname in sys.modules:
            module = sys.modules[fqname]
            if module is None:
                return None
            try:
                spec = module.__spec__
            except AttributeError:
                raise ValueError(f"{fqname}.__spec__ is not set") from None
            else:
                if spec is None:
                    raise ValueError(f"{fqname}.__spec__ is None")
                return fqname

        spec = find_spec(head)
        if not spec:
>           raise ModuleNotFoundError(f"could not find the module {head} to resolve {fqname}", name=head)
E           ModuleNotFoundError: could not find the module Cocoa to resolve Cocoa.NSUserDefaults

.venv\Lib\site-packages\pytest_archon\collect.py:212: ModuleNotFoundError
```

Since it isn't possible to install pyobjc in Windows or Linux, it would be nice to be able to either exclude the module entirely or only try to find the module if it passes the `sys.platform` check.

**Comment by jwbargsten** [comment]
Something like this?

**Comment by jwbargsten** [comment]
Alternatively we could also implement a resolution mapping, e.g.

```python
.check(gaphor, skip_type_checking=True, resolve_map={"Cocoa.NSUserDefaults": "Cocoa"})
```
But I'm not sure if this has added value.

**Comment by danyeaw** [comment]
The resolve whitelist would be OK. I was expecting that when running:

```
.should_not_import("Cocoa")
.should_not_import("Cocoa.NSUserDefaults")
```

That it wouldn't throw an import exception.

**Comment by jwbargsten** [comment]
It fails on the "from Cocoa import NSUserDefaults" construct. Without resolving the module, pytest-archon can not figure out if it is a module or a function.

**Comment by jwbargsten** [comment]
Alternatively we could also just mute the errors and take the "Cocoa.NSUserDefaults" as module name. What do you think, @amolenaar ?

**Comment by amolenaar** [comment]
I think we're interested in if module imports are allowed. 
We're less interested in if the module actually exists. However, that's a requirement if we want to traverse a module.

I think it makes sense to return the module name, I think. The case for platform specific libraries is not uncommon.
Should we raise a warning that a module can't be resolved?

Alternatively users should "mock" the platform specific libraries. That would make it more explicit for the user that a library is platform dependent.

Something like:

```python
    def test_core_packages():
        with platform_specific("darwin", "Cocoa"):
            (
                archrule("gaphor.core does not depend on the rest of the system")
                .match("gaphor.core*")
                .exclude("*.tests.*")
                .may_import(*GAPHOR_CORE)
                .should_not_import("gaphor*")
                .should_not_import(*UI_LIBRARIES)
                .should_not_import("Cocoa")
                .check(gaphor, skip_type_checking=True, resolve_whitelist={"Cocoa.NSUserDefaults"})
            )
```

**Comment by jwbargsten** [comment]
I like it, but I'm also lazy. I therefore suggest to return the full name and issue a warning.

**Comment by amolenaar** [comment]
I think that' best atm.

I had a look at the Sphinx autodoc mock stuff. It's adding a custom module finder, which is something we do not want I suppose.

**Comment by jwbargsten** [comment]
published the fix as v0.0.4

**Comment by danyeaw** [comment]
Thanks for the fix!
